### ISSUE TITLE
Invalid Credentials, Wrong user or password

### ISSUE THREAD

**Issue by jimmykyle** [dlarrick/hass-kumo#109]
I have searched to no avail. Maybe i am doing something silly but I can not get past log in. I have reset my password as well as just updating my password in the Kumo app and the integration will not recognize my credentials.

**Comment by dlarrick** [comment]
Try following the instructions at https://github.com/dlarrick/pykumo#interactive-use -- do your credentials work there? If so, there's something wrong on the Home Assistant side. If not, well, you may be able to pick apart what's going on if you have any Python skills.

**Comment by RandyTidd** [comment]
I'm having the same symptoms as @jimmykyle.  I have no Python skills, but I tried the interactive use anyway. :)

I think my issue is this code does not handle the absence of the address field for a unit, and mine do not have the field: [Pass On Missing Address](https://github.com/dlarrick/pykumo/blob/d388cc85e96489f1554d6f749816977ef8fbb882/pykumo/py_kumo_cloud_account.py#LL67C46-L67C46)

Config flow calls validate_input, which calls account.try_setup, and this fails because it returns False when it finds the unit count is zero and InvalidAuth is raised as a result.

I see that missing ip addresses are handled with custom parsing [here](https://github.com/dlarrick/hass-kumo/blob/6ed591341f3bd1e1d5086d6309a41c1051e8b2e6/custom_components/kumo/config_flow.py#L81).  It seems if we add this handling to _parse_unit in pykumo, my problem might be solved, It looks like there is also an opportunity to remove this custom code and use the pykumo's get_all_units() instead, but maybe there are other reasons this is done this way?

**Comment by RandyTidd** [comment]
I tried modifying /usr/lib/python3.10/site-packages/pykumo/py_kumo_cloud_account.py.  This allowed me to call account.get_outdoor_units() to get the serial number keys in the interactive use way.  Now account.try_setup() now returns True, but that didn't fix my login problem in the integration like I suspected above.  Is there anything else I have to do to get the integration to pick up the code change?  

I can dig into this further, but I'm ignorant on how to get logging working for the integration.  Any tips or instructions for getting that working would be of great help.

**Comment by thomashillyer** [comment]
I’m seeing this issue too. I’m wondering if you still see this?

do you have a mini split system or an indoor outdoor system? I’m wondering if this integration is tightly designed for mini splits only and doesn’t work for my with a PVA indoor air handler. I’ll try and look into it more this weekend for my specific fail

**Comment by RandyTidd** [comment]
I came back to this after a month or so to take another pass at messing with the code, but the issue had gone away and I was able to log in.  The plugin has been working fairly well since.  I did the option to cache things during setup, and I don't think the plugin tries to login again after that (but I don't really know).

I think it's called a mini-split: I have an indoor air handler and an additional head unit in the living room all served by a single outdoor unit.  FWIW, the air handler has the wireless adapter and an MHK2 while the head unit has the wireless adapter and a remote temp. sensor.  I can control both units with the plugin.

**Comment by thomashillyer** [comment]
Hmm ok thanks for clarifying. Mine only has the air handler with wireless adapter and MHK2 (plus kumo cloud adapter). It still won't login when I try to add this integration. Same user and pass works to login to [web kumo](https://app.kumocloud.com/) and the app version.

I did once see a log that said "no json returned" but now I don't see any logs from this integration when I try (and fail) to set it up.

**Comment by thomashillyer** [comment]
@RandyTidd your comment about trying to edit the kumo_cloud_account.py put me onto a fix it seems.

And for logging, I was not able to get great logs from the integration but I found that I could turn on the logging by editing home assistant/configuration.yaml like so:
```
logger:
  default: info
  logs:
    custom_components.hacs: debug # for hacs
    aiogithubapi: debug # also for hacs
    custom_components.kumo: debug
```

**Comment by thomashillyer** [comment]
I tried to add logs to kumo_cloud_setup.py by doing `kumo.log.debug()` instead of the print()s but that didn't seem to work. Might have better luck to do `_LOGGER.debug("foobar")`

**Comment by bbordelon** [comment]
I'm also getting stuck at the initial setup of this integration.  I can log into the iOS app and the app.kumocloud.com website, but using the same credentials in HA results in the "Invalid Credentials, Wrong user or password" error.
I also followed the instructions at https://github.com/dlarrick/pykumo#interactive-use and was able to log in and list my interior units.
Repeated attempts yield the same result, with nothing appearing in the logs.

**Comment by dlarrick** [comment]
If it works in pykumo interactive use the issue has to be in core HomeAssistant, We're not doing anything special, just prompting for the password and storing it in the config.

These credentials appear to be stored by HA in plain text in {your HA directory}/.storage/core.config_entries. Search for `kumo`. Is the username / password listed correctly there?

**Comment by bbordelon** [comment]
It doesn't appear that the config flow gets to the point where it stores the credentials in core.config_entries.  Searched for kumo and nothing's there (yet?).  Tried another config setup and the username/password still aren't in there.

**Comment by dlarrick** [comment]
Thanks for the interactive log. It may well give me a clue of what's going on. Will advise.

**Comment by gregparsons** [comment]
Same authentication error in HA. pykumo login appears to work fine (with an empty command line after password entry). account.get_indoor_units() returns []. I have four units. Anything else I should try? Also, account.try_setup() returns False.

**Comment by KazWolfe** [comment]
Just got a newly installed unit, and can report similar behavior. Doing some digging into interactive mode reveals this is *likely* a problem with the upstream pykumo library:

```
>>> list(account.get_raw_json()[2]['children'][0]['zoneTable'].values())[0]['label']
'Whole Home'
>>> list(account.get_raw_json()[2]['children'][0]['zoneTable'].values())[0]['address']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'address'
```

What is particularly interesting is that I got the library working by changing the `pass` in `parse_unit` to a log line:

```
>>> import pykumo
>>> account = pykumo.KumoCloudAccount.Factory()
Kumo Cloud username: <REDACTED>
Password: 
>>> account.get_indoor_units()
KeyError in parse_unit: 'address'
KeyError in parse_unit: 'address'
KeyError in parse_unit: 'address'
['<REDACTED>']
>>> account.try_setup()
KeyError in parse_unit: 'address'
True
```

Keeping that as `pass` but changing the KeyError in `fetch_if_needed` breaks it again:

```
>>> account = pykumo.KumoCloudAccount.Factory()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'pykumo' is not defined
>>> import pykumo
>>> account = pykumo.KumoCloudAccount.Factory()
Kumo Cloud username: <REDACTED>
Password: 
>>> account.try_setup()
KeyError in _fetch_if_needed: 'serial'
False
```

I can submit a PR to better handle this case and log instead of `pass` if that feels like it can help?

**Comment by bhaggs** [comment]
Running into the same issue attempting to connect to my ducted unit. Was able to authenticate via the pykumo package in a virtual environment but running into the no IP address for the unit. Also noticing what @mt0321 ran into with authenticating behaving inconsistently between shells despite no changes being done to the code. Happy to help debug but unsure where best to start exploring to track down the issue.

**Comment by mt0321** [comment]
@dlarrick I have at least a partial answer. I found a source of nondeterminism between runs that can lead to units randomly failing to parse when a single field (e.g. the IP 'address') is not present. Submitted a pull request to fix: https://github.com/dlarrick/pykumo/pull/47

---

Sharing a couple additional observations in case they help:

1. I was tinkering with pykumo a bunch, and never once have I seen an "address" for my unit in account.get_raw_json(). Is it possible the Kumo Cloud API has stopped returning the local IP address for units entirely (given the number of folks experiencing this issue in recent months)? I know the Kumo app is supposed to be able to work offline too, so not sure how that functionality would still work if this is the case. (I am using an SVZ indoor air handler.)
2. On a newly created zone, the unit's reportedCondition was empty for me until I used the Kumo Cloud app to change the temperature set point for the first time. Just an FYI.
